components/autocomplete/index.js

```diff
@@ -16,104 +17,261 @@ import { keycodes } from '@wordpress/utils';
+				this.replace( range, replacement );
+			}
 		}
 	}
 
 	reset() {
 		this.setState( this.constructor.getInitialState() );
 	}
 
-	onBlur( event ) {
-		// Check if related target is not within, in the case that the user is
-		// selecting an option by button click
-		if ( this.menuNode && ! this.menuNode.contains( event.relatedTarget ) ) {
-			this.reset();
+	// this method is separate so it can be overrided in tests
+	getCursor( container ) {
+		const selection = window.getSelection();
+		if ( selection.isCollapsed && container.contains( selection.anchorNode ) ) {
+			return {
+				node: selection.anchorNode,
+				offset: selection.anchorOffset,
+			};
 		}
+		return null;
 	}
 
-	search( event ) {
-		const { triggerPrefix } = this.props;
-		const { isOpen } = this.state;
-
-		const value = event.target.textContent;
+	// this method is separate so it can be overrided in tests
+	createRange( startNode, startOffset, endNode, endOffset ) {
+		const range = document.createRange();
+		range.setStart( startNode, startOffset );
+		range.setEnd( endNode, endOffset );
+		return range;
+	}
 
-		// Currently search supports only if input's entire value is match
-		let pattern = '^';
+	loadOptions( index ) {
+		this.props.completers[ index ].getOptions().then( ( options ) => {
+			this.setState( {
+				[ 'options_' + index ]: options.map(
+					( option, i ) => ( { ...option, key: index + '_' + i } ) ),
+			} );
+		} );
+	}
 
-		if ( triggerPrefix ) {
-			// Trigger prefix may contain reserved characters of a regular
-			// expression, so escape
-			pattern += escapeStringRegexp( triggerPrefix );
+	findMatch( container, cursor, allCompleters, wasOpen ) {
+		const allowAnything = () => true;
+		let endTextNode;
+		let endIndex;
+		let completers = allCompleters.map( ( completer, idx ) => ( { ...completer, idx } ) );
+		if ( wasOpen ) {
+			// put the open completer at the start so it has priority
+			completers = [
+				wasOpen,
+				...completers.filter( ( completer ) => completer.idx !== wasOpen.idx ),
+			];
 		}
-
-		// Create matching group for the search value
-		pattern += '([^\\s]*)$';
-
-		pattern = new RegExp( pattern );
-
-		let match, search;
-		if ( ( match = value.match( pattern ) ) ) {
-			search = new RegExp( escapeStringRegexp( match[ 1 ] ), 'i' );
+		// search backwards to find the first preceeding space or non-text node.
+		if ( isTextNode( cursor.node ) ) { // TEXT node
+			endTextNode = cursor.node;
+			endIndex = cursor.offset;
+		} else if ( cursor.offset === 0 ) {
+			endTextNode = onlyTextNode( descendFirst( cursor.node ) );
+			endIndex = 0;
+		} else {
+			endTextNode = onlyTextNode( descendLast( cursor.node.childNodes[ cursor.offset - 1 ] ) );
+			endIndex = endTextNode ? endTextNode.nodeValue.length : 0;
 		}
-
-		if ( isOpen ) {
-			if ( match ) {
-				// Reset selection to initial when search changes
-				this.setState( {
-					selectedIndex: 0,
-					search,
-				} );
+		if ( endTextNode === null ) {
```

Comment:
Can we expect this to occur often? Seeing as `completers` is not used to this point, we could avoid mapping `allCompleters` until _after_ we finish checking early returns, to avoid wasted logic run on every keystroke in editable.

Reply:
I have moved the calculation of `completers`.